anomaly-benchmarking/script.whizzml

```diff
@@ -0,0 +1,277 @@
+    (loop (queue [0] nodes (color-a-node old-nodes 0 "normal"))
+      (if (empty? queue)
+          (make-map ["nodes" "edges"] [nodes edges])
+          (let (node (head queue)
+                rest (tail queue)
+                node-color (mapify nodes "id" "color")
+                neighbors (filter (lambda (x) (uncolored? x node-color)) 
+                                  (find-neighbors node span-tree))
+                colored-nodes (color-lots nodes neighbors 
+                                         (new-color (node-color (str node))))
+                new-queue (concat rest neighbors))
+            (recur new-queue colored-nodes))))))
+
+;; Creating the new dataset
+
+;; Given a field id and a map of class names to colors, returns a
+;; flatline expression that returns a color based on the class in the
+;; given field.
+
+(define (color-filter field-id color-map)
+  (let (classes (keys color-map)
+        one-line (lambda (class color) 
+                   (flatline "(matches? (field {{field-id}}) {{class}}) {{color}}"))
+        conditions (reduce (lambda (x y) 
+                       (join " " [x (one-line y (color-map y))])) 
+                     "" (keys color-map)))
+    (join " " ["(cond" conditions ")"])))
+
+;; Given a dataset with a multivariate objective field, returns an
+;; extended dataset with a new field "Is Anomaly?", which can be
+;; either "normal" or "anomalous", determined by the two-colored graph
+;; of the confusion matrix
+
+(define (multi-to-binary ds)
+  (let (my-eval (evaluate-model-on-dataset ds)
+       my-graph (make-graph my-eval)
+       my-span-tree (max-span-tree my-graph)
+       colored-graph (two-color my-graph my-span-tree)
+       field-id (get (fetch ds) "objective_field")
+       color-map (mapify (get colored-graph "nodes") "name" "color")
+       filter (color-filter (get field-id "id") color-map))
+    (create-dataset {"origin_dataset" ds
+                     "new_fields" [{
+                       "field" filter
+                       "name" "Is Anomaly?"}]})))
+
+;; Numeric-to-Binary
+
+;; Given a field id for a numeric field and a median value, returns a
+;; flatline command that categorizes datapoints above the median as
+;; "anomalous" and below as "normal"
+
+(define (median-filter id median)
+  (flatline "(if (> (field {{id}}) {{median}}) \"anomalous\" \"normal\")"))
+
+;; Given a dataset with a numeric objective field, returns an extended
+;; dataset with a new field "binary" with values "anomalous" and
+;; "normal" based on whether the value of the objective field is above
+;; or below the median
+
+(define (numeric-to-binary ds)
+  (let (my-ds (fetch ds)
+        field-id (get-in my-ds ["objective_field" "id"])
+        median (get-in my-ds ["fields" field-id "summary" "median"]))
+    (create-dataset {"origin_dataset" ds
+                     "new_fields" [{
+                       "field" (median-filter field-id median)
```

Comment:
and here similarly - `"field"` should be on the same line as the opening brace, and I usually bind things so that I can get all key-value pairs on one line.